workspace/__ardulib__/WS2812_NoBufffer/WS2812.cpp

```diff
@@ -0,0 +1,152 @@
+            ".endr \n\t" ::                           \
+                [port] "I"(_SFR_IO_ADDR(_PORT)),      \
+            [bit] "I"(_BIT),                          \
+            [onCycles] "I"(NS_TO_CYCLES(_ON) - 2),    \
+            [offCycles] "I"(NS_TO_CYCLES(_OFF) - 2)); \
+    }
+
+#define SEND_BITS(_ON, _OFF, _PORT) \
+    SEND_BIT(_ON, _OFF, _PORT, 0);  \
+    SEND_BIT(_ON, _OFF, _PORT, 1);  \
+    SEND_BIT(_ON, _OFF, _PORT, 2);  \
+    SEND_BIT(_ON, _OFF, _PORT, 3);  \
+    SEND_BIT(_ON, _OFF, _PORT, 4);  \
+    SEND_BIT(_ON, _OFF, _PORT, 5);
+
+// Actually send a bit to the string. We must to drop to asm to enusre that the complier does
+// not reorder things and make it so the delay happens in the wrong place.
+void WS2812::sendBit(bool bitVal) {
+    //    _bit = 5;
+    if (bitVal) {
+        // 0 bit
+#ifdef PORTA
+        SEND_BITS(T1H, T1L, PORTA);
+#endif
+#ifdef PORTB
+        SEND_BITS(T1H, T1L, PORTB);
+#endif
+#ifdef PORTC
+        SEND_BITS(T1H, T1L, PORTC);
+#endif
+#ifdef PORTD
+        SEND_BITS(T1H, T1L, PORTD);
+#endif
+#ifdef PORTE
+        SEND_BITS(T1H, T1L, PORTE);
+#endif
+#ifdef PORTF
+        SEND_BITS(T1H, T1L, PORTF);
+#endif
+#ifdef PORTG
+        SEND_BITS(T1H, T1L, PORTG);
+#endif
+    } else {
+        // 1 bit
+#ifdef PORTA
+        SEND_BITS(T0H, T0L, PORTA);
+#endif
+#ifdef PORTB
+        SEND_BITS(T0H, T0L, PORTB);
+#endif
+#ifdef PORTC
+        SEND_BITS(T0H, T0L, PORTC);
+#endif
+#ifdef PORTD
+        SEND_BITS(T0H, T0L, PORTD);
+#endif
+#ifdef PORTE
+        SEND_BITS(T0H, T0L, PORTE);
+#endif
+#ifdef PORTF
+        SEND_BITS(T0H, T0L, PORTF);
+#endif
+#ifdef PORTG
+        SEND_BITS(T0H, T0L, PORTG);
+#endif
+    }
+}
+
+#if defined(ESP8266)
+// ESP8266 show() is external to enforce ICACHE_RAM_ATTR execution
+extern "C" void ICACHE_RAM_ATTR espSendByte(uint8_t pin, uint8_t byte);
+#elif defined(ESP32)
+extern "C" void espSendByte(uint8_t pin, uint8_t byte);
+#endif // ESP8266
+
+void WS2812::sendByte(uint8_t byte) {
+#if defined(ESP8266) || defined(ESP32)
+    // ESP8266 show() is external to enforce ICACHE_RAM_ATTR execution
+    espSendByte(_bit, byte);
+#else
+    for (uint8_t bit = 0; bit < 8; bit++) {
+        sendBit(bitRead(byte, 7)); // Neopixel wants bit in highest-to-lowest order
+            // so send highest bit (bit #7 in an 8-bit byte since they start at 0)
+        byte <<= 1; // and then shift left so bit 6 moves into 7, 5 moves into 6, etc
+    }
+#endif
+}
+
+void WS2812::sendPixel(uint8_t r, uint8_t g, uint8_t b) {
+    // Neopixel wants colors in green then red then blue order
+    sendByte(g);
+    sendByte(r);
+    sendByte(b);
+}
+
+// Just wait long enough without sending any bots to cause the pixels to latch and display the last sent frame
+void WS2812::show() {
+    _delay_us((RES / 1000UL) + 1); // Round up since the delay must be _at_least_ this long (too short might not work, too long not a problem)
+}
+
+// Display a single color on the whole string
```

Comment:
```suggestion
// Display a single color on the whole strip
```

Find and replace everywhere as well. String is the “Hello”, the LEDs is a _strip_.